consumer/src/main/kotlin/streams/kafka/KafkaEventSink.kt

```diff
@@ -113,15 +148,172 @@ class KafkaEventConsumer(private val consumer: KafkaConsumer<String, ByteArray>,
         StreamsUtils.ignoreExceptions({ consumer.close() }, UninitializedPropertyAccessException::class.java)
     }
 
-    override fun read(): Map<String, List<Any>>? {
-        val records = consumer.poll(config.sinkPollingInterval)
-        if (records != null && !records.isEmpty) {
-            return records
-                    .map {
-                        it.topic()!! to JSONUtils.readValue<Any>(it.value())
+    private fun readSimple(action: (String, List<Any>) -> Unit) {
+        val records = consumer.poll(0)
+        if (!records.isEmpty) {
+            try {
+                this.topics.forEach { topic ->
+                    action(topic, records.records(topic).map { JSONUtils.readValue<Any>(it.value()) })
+                }
+            } catch (e: Exception) {
+                // TODO add dead letter queue
+            }
+        }
+    }
+
+    private fun readFromPartition(config: KafkaTopicConfig, action: (String, List<Any>) -> Unit) {
+        setSeek(config.topicPartitionsMap)
+        val records = consumer.poll(0)
+        val consumerRecordsMap = toConsumerRecordsMap(config.topicPartitionsMap, records)
+        if (consumerRecordsMap.isNotEmpty()) {
+            try {
+                consumerRecordsMap.forEach { action(it.key.topic(), it.value.map { JSONUtils.readValue<Any>(it.value()) }) }
+            } catch (e: Exception) {
+                // TODO add dead letter queue
+            }
+        }
+    }
+
+    override fun read(topicConfig: Map<String, Any>, action: (String, List<Any>) -> Unit) {
+        val kafkaTopicConfig = KafkaTopicConfig.fromMap(topicConfig)
+        if (kafkaTopicConfig.topicPartitionsMap.isEmpty()) {
+            readSimple(action)
+        } else {
+            readFromPartition(kafkaTopicConfig, action)
+        }
+    }
+
+    fun toConsumerRecordsMap(topicPartitionsMap: Map<TopicPartition, Long>,
+                             records: ConsumerRecords<String, ByteArray>)
+            : Map<TopicPartition, List<ConsumerRecord<String, ByteArray>>> = topicPartitionsMap
+                .mapValues {
+                    records.records(it.key)
```

Comment:
naming would be good here, I think you can deconstruct pairs as `.mapValues { (topic, partition) -> }`

Reply:
https://play.kotlinlang.org/#eyJ2ZXJzaW9uIjoiMS4zLVJDIiwicGxhdGZvcm0iOiJqYXZhIiwiYXJncyI6IiIsIm5vbmVNYXJrZXJzIjp0cnVlLCJ0aGVtZSI6ImlkZWEiLCJmb2xkZWRCdXR0b24iOnRydWUsInJlYWRPbmx5IjpmYWxzZSwiY29kZSI6ImZ1biBtYWluKCkgeyBcbnZhbCBhID0gbGlzdE9mKFwiYVwiIHRvIDIsXCJiXCIgdG8gMykudG9NYXAoKVxuXG52YWwgYiA9IGEubWFwVmFsdWVzeyAoa2V5LCB2YWx1ZSkgLT4gXG4gICAgdmFsdWUgKzFcbn1cbnByaW50bG4oYilcblxufSJ9